### DISCUSSION TITLE
8 & 10 bits conversion

### DISCUSSION THREAD

**Discussion by ghost** [ImageMagick/ImageMagick#2357] [Help] [answered]
Hi,
I discovered a strange behavior with the conversion of 10-bit images.
I have an unformatted image containing 10-bits pixel gray values of 320 x 240 pixels. As the input file is not formatted, I use the following command to generate an HDR file: 

convert -size 320x240 -depth 16 -endian MSB gray:image.raw image.hdr

I have the feeling that the image is trated like a 8-bit image. Sometimes I have a luminance rollower (like if the MSB is not considered). I have the same behavior if I want to generate png or jpg files.
![image](https://user-images.githubusercontent.com/5145938/89185112-fe37be00-d599-11ea-89ab-26a28dd3a50c.jpg)

I attach an image (not at the HDR format) just to emphasize the problem (bottom - left). The same command with -depth 8 works perfectly. The analysis of the raw file emphasize that the MSB is not considered for the conversion.
Any idea or suggestion?
Edo

**Comment by snibgo** [comment]
What version of IM, on what platform?

IM generally assumes files are unsigned, but your file may be signed, giving the wraparound problem.

You say your input is 10 bits per pixel, grayscale. Does that mean the file stores each pixel in 2 bytes (16 bits) with the top or bottom 4 bits unused and set to zero? Or are the pixels packed so 10 bytes (80 bits) stores 8 pixels?

Please link to a sample input.

**Reply by ghost** [reply]
See my next post.
Anyway, a rapid analysis of the image.raw confirms that something is wrong with the reading of the image.
I have isolated a couple of pixels close to the area that shows the problem.
In green the right one 0x00DE & 0x00E3 (I swapped MSB & LSB due to the Endianess). Here, the luminance is still represented in 8-bits. I red, we can see that the pixels values are 0x011D & 0x011D. So, the red pixels should be brighter than the green one. Now, it is the opposite!! The reason I can see, is that the conversion does not take in account the MSB. In this case, the red pixels are considered as 0x001D & 0x001D .... 0x001D is lower than 0x00DE & 0x00E3, so, we have the wrong dark region.

How to overtake this?

<img width="728" alt="Screen Shot 2020-08-04 at 00 23 57" src="https://user-images.githubusercontent.com/5145938/89233604-3a930a80-d5ea-11ea-83ca-b3a91d542c91.png">

**Comment by ghost** [comment]
Hi Snibgo,
Thank you for your answer.
The IM I use is the latest available on Macport, (ImageMagick-6.9.11-25).
The used platform is macOS 10.15.6

The data of the raw image are unsigned. I use 2 bytes for representing 10-bits. The unused bits are set to zero.
The coding is this one:

Bit15.........................................Bit0
MSB......................LSB
0 0 0 0 0 0 v v .....v v v v v v v v

Of course, we need to be careful due to the indianess representation.
In the raw file, the organisation is this one:

LSB0 MSB0 LSB1 MSB1 LSB2 MSB2 .... LSBn MSBn

For this reason in my command I use  -endian MSB in order to rearrange the bytes.

Attached one raw image.raw and its conversion result image.hdr

Best, Edo
[images.zip](https://github.com/ImageMagick/ImageMagick/files/5016887/images.zip)

**Comment by snibgo** [comment, marked as answer]
> (I swapped MSB & LSB due to the Endianess)

Huh?

What command are you currently using?

You say:
> 0 0 0 0 0 0 v v .....v v v v v v v v
LSB0 MSB0 LSB1 MSB1 LSB2 MSB2 .... LSBn MSBn
For this reason in my command I use -endian MSB

That is wrong. The lowest address is the least significant byte, so use "-endian LSB".

```
magick -size 320x240 -depth 16 -endian LSB gray:image.raw +endian x.png
```
As only the bottom 10 bits out of 16 are used, the maximum value is 1023. Assuming you use IM Q16, we can multiply the result by 65535/1023:
```
magick -size 320x240 -depth 16 -endian LSB gray:image.raw +endian -evaluate multiply %[fx:65535/1023] x.png
```

**Reply by ghost** [reply]
Thank you snibgo,
That is a bit cryptic but it works :-) 

In particular is not clear the usage of the +endian ...

Just a small modification (adding the '' around %[fx:...] )

`magick -size 320x240 -depth 16 -endian LSB gray:image.raw +endian -evaluate multiply '%[fx:65535/1023]' x.hdr`

otherwise the terminal complains.
